src/Chainweb/Mempool/InMem.hs

```diff
@@ -201,60 +206,106 @@ markValidatedInMem :: MVar (InMemoryMempoolData t)
 insertCheckInMem cfg lock txs = do
     now <- getCurrentTimeIntegral
     badmap <- withMVarMasked lock $ readIORef . _inmemBadMap
-    let hashes = V.map hasher txs
-    let txhashes = V.zip hashes txs
-    let checks = [ (InsertErrorOversized, sizeOK)
-                 , (InsertErrorInvalidTime, ttlCheck now)
-                 , (InsertErrorBadlisted, notInBadMap badmap)
-                 ]
-    let out1 = V.map (\tx -> runPreChecks checks tx) txhashes
-    out2 <- _inmemPreInsertCheck cfg txs
-    return $! V.zip hashes (V.zipWith (<|>) out1 out2)
 
+    -- We hash the tx here and pass it around around to avoid needing to repeat
+    -- the hashing effort.
+    let withHashes :: Either (TransactionHash, InsertError) (Vector (TransactionHash, t))
+        withHashes = for txs $ \tx ->
+          let !h = hasher tx
+          in bimap (h,) (h,) $ validateOne cfg badmap now tx h
+
+    case withHashes of
+        Left _ -> pure $ void withHashes
+        Right r -> void . sequenceA <$> _inmemPreInsertBatchChecks cfg r
+  where
+    hasher :: t -> TransactionHash
+    hasher = txHasher (_inmemTxCfg cfg)
+
+-- | Validation: Confirm the validity of some single transaction @t@.
+--
+validateOne
+    :: forall t a
+    .  InMemConfig t
+    -> HashMap TransactionHash a
+    -> Time Micros
+    -> t
+    -> TransactionHash
+    -> Either InsertError t
+validateOne cfg badmap (Time (TimeSpan now)) t h =
+    sizeOK
+    >> ttlCheck
+    >> notInBadMap
+    >> _inmemPreInsertPureChecks cfg t
   where
-    notInBadMap badmap (h, _) = not (HashMap.member h badmap)
+    txcfg :: TransactionConfig t
     txcfg = _inmemTxCfg cfg
-    hasher = txHasher txcfg
-    getSize = txGasLimit txcfg
-    maxSize = _inmemTxBlockSizeLimit cfg
-    sizeOK (_, tx) = getSize tx <= maxSize
+
+    sizeOK :: Either InsertError ()
+    sizeOK = ebool_ InsertErrorOversized (getSize t <= maxSize)
+      where
+        getSize = txGasLimit txcfg
+        maxSize = _inmemTxBlockSizeLimit cfg
+
     -- prop_tx_ttl_arrival
-    ttlCheck (Time (TimeSpan now)) (_, tx) =
-      case txMetadata txcfg tx of
-        TransactionMetadata (Time (TimeSpan creationTime)) (Time (TimeSpan expiryTime)) ->
-            creationTime < now && now < expiryTime && creationTime < expiryTime
+    ttlCheck :: Either InsertError ()
+    ttlCheck = ebool_ InsertErrorInvalidTime (ct < now && now < et && ct < et)
+      where
+        TransactionMetadata (Time (TimeSpan ct)) (Time (TimeSpan et)) = txMetadata txcfg t
 
-    runPreChecks [] _ = Nothing
-    runPreChecks ((reason, chk):chks) tx =
-        if chk tx
-          then runPreChecks chks tx
-          else Just reason
+    notInBadMap :: Either InsertError ()
+    notInBadMap = maybe (Right ()) (const $ Left InsertErrorBadlisted) $ HM.lookup h badmap
+
+-- | Validation: Similar to `insertCheckInMem`, but does not short circuit.
+-- Instead, bad transactions are filtered out and the successful ones are kept.
+--
+insertCheckInMem'
```

Comment:
This variation is used below in `insertInMem`, who doesn't want to fail fast, and instead wants to insert all the TXs that passed anyway, even if there were some failures in the group.